From f0ec9a1726a66278249f58a46c2b4f9394b32eff Mon Sep 17 00:00:00 2001
From: Ross Lagerwall <ross.lagerwall@citrix.com>
Date: Fri, 20 Mar 2015 14:32:04 +0000
Subject: [PATCH 02/26] efi: Ensure incorrectly typed runtime services get
 mapped

Some firmware implementations do not correctly mark memory needed for runtime
services, not setting the EFI_MEMORY_RUNTIME bit and giving it a type
EfiReservedMemoryType.  Even though EfiReservedMemoryType is not supposed to be
used by the firmware, map these regions so that runtime services work.

The failing firmware implementations were:
    Product Name: PowerEdge R720
    Vendor: Dell Inc.
    Version: 2.1.2
    Release Date: 09/19/2013
    BIOS Revision: 2.1

    Product Name: PowerEdge R320
    Vendor: Dell Inc.
    Version: 1.5.2
    Release Date: 03/11/2013
    BIOS Revision: 1.5

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
---
 xen/common/efi/boot.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index d78a315e56e0..4fb755bc4e43 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -156,6 +156,7 @@ static SIMPLE_TEXT_OUTPUT_INTERFACE *__initdata StdErr;
 
 static UINT32 __initdata mdesc_ver;
 static bool __initdata map_bs;
+static bool __initdata map_res = true;
 
 static struct file __initdata cfg;
 static struct file __initdata kernel;
@@ -1717,6 +1718,11 @@ void __init efi_init_memory(void)
                 if ( !map_bs )
                     continue;
                 break;
+
+            case EfiReservedMemoryType:
+                if ( !map_res )
+                    continue;
+                break;
             }
         }
 
-- 
2.37.3

